provider/azure.go

```diff
@@ -0,0 +1,456 @@
+	}
+	return fmt.Sprintf("%s.%s", recordName, zoneName)
+}
+
+func extractTarget(recordType string, recordSet *dns.RecordSet) string {
+	properties := recordSet.RecordSetProperties
+	if properties == nil {
+		return ""
+	}
+
+	switch dns.RecordType(recordType) {
+	case dns.A:
+		aRecords := properties.ARecords
+		if aRecords != nil && len(*aRecords) > 0 && (*aRecords)[0].Ipv4Address != nil {
+			return *(*aRecords)[0].Ipv4Address
+		}
+	case dns.CNAME:
+		cnameRecord := properties.CnameRecord
+		if cnameRecord != nil && cnameRecord.Cname != nil {
+			return *cnameRecord.Cname
+		}
+	case dns.TXT:
+		txtRecords := properties.TxtRecords
+		if txtRecords != nil && len(*txtRecords) > 0 && (*txtRecords)[0].Value != nil {
+			values := (*txtRecords)[0].Value
+			if values != nil && len(*values) > 0 {
+				return (*values)[0]
+			}
+		}
+	}
+	return ""
+}
+
+type azureChangeMap map[*dns.Zone][]*endpoint.Endpoint
+
+func mapAzureChanges(zones []dns.Zone, changes *plan.Changes) (azureChangeMap, azureChangeMap) {
+	ignored := map[string]bool{}
+	deleted := azureChangeMap{}
+	updated := azureChangeMap{}
+
+	mapChange := func(changeMap azureChangeMap, change *endpoint.Endpoint) {
+		zone := findAzureZone(zones, change.DNSName)
+		if zone == nil {
+			if _, ok := ignored[change.DNSName]; !ok {
+				ignored[change.DNSName] = true
+				log.Infof("Ignoring changes to '%s' because a suitable Azure DNS zone was not found.", change.DNSName)
+			}
+			return
+		}
+		// Change unspecified record types to address
+		if change.RecordType == "" {
+			change.RecordType = "A"
+		}
+		changeMap[zone] = append(changeMap[zone], change)
+	}
+
+	for _, change := range changes.Delete {
+		mapChange(deleted, change)
+	}
+
+	for _, change := range changes.UpdateOld {
+		mapChange(deleted, change)
+	}
+
+	for _, change := range changes.Create {
+		mapChange(updated, change)
+	}
+
+	for _, change := range changes.UpdateNew {
+		mapChange(updated, change)
+	}
+	return deleted, updated
+}
+
+func findAzureZone(zones []dns.Zone, name string) *dns.Zone {
+	var result *dns.Zone
+
+	// Go through every zone looking for the longest name (i.e. most specific) as a matching suffix
+	for idx := range zones {
+		zone := &zones[idx]
+		if strings.HasSuffix(name, *zone.Name) {
+			if result == nil || len(*zone.Name) > len(*result.Name) {
+				result = zone
+			}
+		}
+	}
+	return result
+}
+
+func (p *AzureProvider) deleteAzureRecords(deleted azureChangeMap) {
+	// Delete records first
+	for zone, endpoints := range deleted {
+		for _, endpoint := range endpoints {
+			name := recordSetNameForZone(zone, endpoint)
+			if p.dryRun {
+				log.Infof("Would delete %s record named '%s' for Azure DNS zone '%s'.", endpoint.RecordType, name, *zone.Name)
+			} else {
+				log.Infof("Deleting %s record named '%s' for Azure DNS zone '%s'.", endpoint.RecordType, name, *zone.Name)
+				if _, err := p.recordsClient.Delete(p.resourceGroup, *zone.Name, name, dns.RecordType(endpoint.RecordType), ""); err != nil {
```

Comment:
does Azure support batch delete/update ? It is potentially dangerous to have only portion of the records changed, for example only txt record is deleted but the A record is preserved